src/daft-io/src/local.rs

```diff
@@ -84,49 +98,83 @@ pub struct LocalFile {
-            }))
-        } else {
-            return Err(Error::InvalidFilePath {
-                path: uri.to_string(),
-            }
-            .into());
-        }
+        const LOCAL_PROTOCOL: &str = "file://";
+        let path = uri.strip_prefix(LOCAL_PROTOCOL).unwrap_or(uri);
+        Ok(GetResult::File(LocalFile {
+            path: path.into(),
+            range,
+        }))
     }
 
     async fn get_size(&self, uri: &str) -> super::Result<usize> {
-        const TO_STRIP: &str = "file://";
-        if let Some(p) = uri.strip_prefix(TO_STRIP) {
-            let path = std::path::Path::new(p);
-            let file = tokio::fs::File::open(path)
-                .await
-                .context(UnableToOpenFileSnafu {
-                    path: path.to_string_lossy(),
-                })?;
-            let metadata = file.metadata().await.context(UnableToOpenFileSnafu {
-                path: path.to_string_lossy(),
+        const LOCAL_PROTOCOL: &str = "file://";
+        let path = uri.strip_prefix(LOCAL_PROTOCOL).unwrap_or(uri);
+        let meta = tokio::fs::metadata(path)
+            .await
+            .context(UnableToFetchFileMetadataSnafu {
+                path: path.to_string(),
             })?;
-            return Ok(metadata.len() as usize);
-        } else {
-            return Err(Error::InvalidFilePath {
-                path: uri.to_string(),
-            }
-            .into());
-        }
+        Ok(meta.len() as usize)
     }
 
     async fn ls(
         &self,
-        _path: &str,
+        path: &str,
         _delimiter: Option<&str>,
         _continuation_token: Option<&str>,
     ) -> super::Result<LSResult> {
-        unimplemented!("local ls");
+        const LOCAL_PROTOCOL: &str = "file://";
+        let path = path.strip_prefix(LOCAL_PROTOCOL).unwrap_or(path);
+        let meta = tokio::fs::metadata(path)
+            .await
+            .context(UnableToFetchFileMetadataSnafu {
+                path: path.to_string(),
+            })?;
+        if meta.file_type().is_file() {
+            // Provided path points to a file, so only return that file.
+            return Ok(LSResult {
+                files: vec![FileMetadata {
+                    filepath: path.into(),
+                    size: Some(meta.len()),
+                    filetype: object_io::FileType::File,
+                }],
+                continuation_token: None,
+            });
+        }
+        // NOTE(Clark): read_dir follows symbolic links, so no special handling is needed there.
+        let dir_entries =
+            tokio::fs::read_dir(path)
+                .await
+                .context(UnableToFetchDirectoryEntriesSnafu {
+                    path: path.to_string(),
+                })?;
+        let mut dir_stream = tokio_stream::wrappers::ReadDirStream::new(dir_entries);
+        let size = dir_stream.size_hint().1.unwrap_or(0);
+        let mut files = Vec::with_capacity(size);
+        while let Some(entry) =
```

Comment:
You may actually want to override [`iter_dir` ](https://github.com/Eventual-Inc/Daft/blob/main/src/daft-io/src/object_io.rs#L88) and return a Stream<FileMetadata> there and then call it from ls rather than collect the stream here and return it as a vec.

Reply:
Any particular reason we don't have `ls` return a stream as well?

Reply:
Main reason is that S3, GCS, HTTP return materialized results, (typically 1000 at a time, or all a single page worth for http). `iter_dir` is the stream version leveraging the continuation token calling ls multiple times.